Improved ML button styling

AIC-BV opened the pull request:
Not perfect, but much better than it actually was. Feel free to edit however you see fit!
The button is more obvious. Users will see more easily that they have to translate the corresponding field.

mjauvin commented:
@AIC-BV there is a problem with markdown editor:
![image](https://github.com/wintercms/wn-translate-plugin/assets/2013630/3982c659-a9f6-4284-b06d-4ca50a349b7b)

It is hiding the "preview" (eye) icon:
![image](https://github.com/wintercms/wn-translate-plugin/assets/2013630/971245a5-d814-43c6-a74d-7cd4fb64d675)

AIC-BV commented:
@mjauvin Must've overlooked it when converting the CSS to LESS

mjauvin commented:
In the `.ml-btn` class, can you remove `height:...` and add `bottom: 1px;` and `box-shadow: none;` ?

AIC-BV commented:
Didn't know adding the bottom would actually set its height correctly!
Better now @mjauvin?

mjauvin commented:
Really nice, one more thing:
```diff
diff --git a/assets/less/multilingual.less b/assets/less/multilingual.less
index db20a4a..82465ef 100644
--- a/assets/less/multilingual.less
+++ b/assets/less/multilingual.less
@@ -65,7 +65,8 @@
 
     &.field-multilingual-markdowneditor {
         .ml-btn {
-            top: 41px;
+            right: 76px;
+            background-color: transparent;
         }
 
         .ml-dropdown-menu {
```

mjauvin commented:
Would be nice to move the ml-btn on the RichEditor as well, but it seems more difficult.

AIC-BV commented:
> Really nice, one more thing:
> 
> ```diff
> diff --git a/assets/less/multilingual.less b/assets/less/multilingual.less
> index db20a4a..a174cc8 100644
> --- a/assets/less/multilingual.less
> +++ b/assets/less/multilingual.less
> @@ -65,12 +65,13 @@
>             
>      &.field-multilingual-markdowneditor {
>          .ml-btn {
> -            top: 41px;
> +            right: 76px;
> +            background-color: transparent;
>          }
>  
>          .ml-dropdown-menu {
> -            top: 74px;
> -            right: 1px;
> +            top: 36px;
> +            right: 76px;
>          }
>      }
> ```
> 
> ![image](https://user-images.githubusercontent.com/2013630/272315195-bdaa67c2-611d-431b-9ca4-d8ac95fc93e5.png)

I added this!

mjauvin commented:
@AIC-BV can you paste a screenshot of what the repeater ML-button looks like here?

AIC-BV commented:
It was like this
<img width="643" alt="image" src="https://github.com/wintercms/wn-translate-plugin/assets/89913092/a20bbf8d-b222-4127-87aa-649d2c793e8f">

but some changes to other things made it this
<img width="635" alt="image" src="https://github.com/wintercms/wn-translate-plugin/assets/89913092/d03de1f6-d4ea-456e-b186-be9365b65538">

Can make it back to the first screenshot by adding
`bottom: auto;`

What do you think?

mjauvin commented:
Yes, do what's required!

AIC-BV commented:
@mjauvin should be done

mjauvin commented:
Dam, doesn't work on mobile
![Screenshot_20231004-211533](https://github.com/wintercms/wn-translate-plugin/assets/2013630/64d125b6-d08d-4e88-a8fb-82f4f0e5e816)

mjauvin commented:
Returns markdown editor ml-button to previous location and add margin to text input box so that we don't end-up under the ml-button

```diff
diff --git a/assets/less/multilingual.less b/assets/less/multilingual.less
index 237ba1a..aaa1dfa 100644
--- a/assets/less/multilingual.less
+++ b/assets/less/multilingual.less
@@ -69,14 +69,17 @@
     }
 
     &.field-multilingual-markdowneditor {
+        .editor-write {
+            margin-right: 2rem;
+        }
         .ml-btn {
-            right: 76px;
-            background-color: transparent;
+            top: 41px;
+            border-radius: 0;
+            border-bottom-left-radius: .375rem;
         }
 
         .ml-dropdown-menu {
-            top: 36px;
-            right: 76px;
+            top: 80px;
         }
     }
```

AIC-BV commented:
@mjauvin Added your changes, and fixed ML repeater and ML richeditor (switched them around by accident 🤪)
I also added the `.editor-write` code to the richeditor to prevent the same issue.

There is however one thing left.
The richeditor button is getting its top set via javascript but I can't find where. I don't know why it does that, but I think it is unnecessary and should be removed?

mjauvin commented:
We're getting closer (can't find the `top` property being added dynamically either), but try this for the margins instead:
```diff
diff --git a/assets/less/multilingual.less b/assets/less/multilingual.less
index f7c2c2e..5c81765 100644
--- a/assets/less/multilingual.less
+++ b/assets/less/multilingual.less
@@ -62,8 +62,8 @@
     }
 
     &.field-multilingual-richeditor {
-        .editor-write {
-            margin-right: 2rem;
+        .fr-wrapper {
+            padding-right: 23px;
         }
         
         .ml-btn {
```

AIC-BV commented:
👍

mjauvin commented:
Did you find where the richeditor's top property gets dynamically modified @AIC-BV ?

I'm ready to merge this when that gets fixed.

AIC-BV commented:
> Did you find where the richeditor's top property gets dynamically modified @AIC-BV ?
https://github.com/wintercms/wn-translate-plugin/blob/f5f7aa1626aa42d3fbd5430ae0c04cd34ca5ffdd/formwidgets/mlricheditor/assets/js/mlricheditor.js#L95

AIC-BV commented:
@mjauvin I tested this and my comment above must be the case.
I guess the markdown editor has this issue too?

So the JavaScript should stay, but for the new styling we can remove the 6px.

<img width="325" alt="image" src="https://github.com/wintercms/wn-translate-plugin/assets/89913092/88dfd4b1-753b-4287-9268-a41b06b47595">

AIC-BV commented:
@mjauvin The markdown gets a horizontal scrollbar on the toolbar, so that's why there is no issue for the markdown.

mjauvin commented:
@AIC-BV maybe we should modify the js code to only add those then the toolbar height is > 0 ?

AIC-BV commented:
@mjauvin Like this?

mjauvin commented:
Looks awesome @AIC-BV but there is still a case where it needs work... when the markdown/richeditor have a scrollbar visible, it breaks like in the screenshots below:
![image](https://github.com/wintercms/wn-translate-plugin/assets/2013630/f01278e3-7285-49cb-81f0-ff532e4b5e52)
![image](https://github.com/wintercms/wn-translate-plugin/assets/2013630/bacc7af1-0ca8-4745-996a-4a7693bdcfd7)

AIC-BV commented:
@mjauvin removing the
```
.fr-wrapper {
    padding-right: 23px;
}
```
puts the scrollbar in the right place. Why did you add the padding? I don't think it is necessary?
<img width="369" alt="image" src="https://github.com/wintercms/wn-translate-plugin/assets/89913092/c362c419-2fa1-4e64-9aaa-72f75bbc260a">

Regardless, do you think it is an issue that the language button is over the scrollbar itself?
Because in that case, the .ml-btn would need `right: +-17px` which could be done through javascript since its already there.
```
if (element.scrollHeight > element.offsetHeight) button.style.right = '17px'
```
<img width="272" alt="image" src="https://github.com/wintercms/wn-translate-plugin/assets/89913092/ae5d7518-09a6-4b7d-a783-c1e2a0035967">

mjauvin commented:
Ok, let's do this with JavaScript.

AIC-BV reviewed (commented):
* mlricheditor

AIC-BV commented:
@mjauvin can you check? 😄

AIC-BV commented:
@mjauvin Only 'problem' I currently have is that the positions don't get updated while inputting the content.
It only gets the correct offset on page load

mjauvin commented:
> @mjauvin Only 'problem' I currently have is that the positions don't get updated while inputting the content. It only gets the correct offset on page load. To avoid the button going over the content/scrollbar, the offset needs to be checked every character input...

Yeah, not great.

I tried something else and I think it's a nice compromise... making the width of the ml-btn a bit wider and more padding on the right so its text shift to the left.

> Also, Mac's don't have a visible scrollbar? How will this offset look on a mac?

The above solution would fix this as well, I guess.

mjauvin commented:
This change is acceptable I think:
```diff
--- a/assets/less/multilingual.less
+++ b/assets/less/multilingual.less
@@ -74,12 +74,10 @@
     }
 
     &.field-multilingual-markdowneditor {
-        .editor-write {
-            margin-right: 2rem;
-        }
-
         .ml-btn {
             top: 41px;
+            width: 3rem;
+            padding-inline: 0 8px;
             border-radius: 0;
             border-bottom-left-radius: .375rem;
         }
```

We need to do the equivalent for the richeditor.

![Screenshot from 2023-10-17 09-15-58](https://github.com/wintercms/wn-translate-plugin/assets/2013630/7887c88b-a8f4-433f-8696-fccdfbed24f8)

mjauvin commented:
That works for the richeditor, but the scrollbar is hidden:
```diff
diff --git a/assets/less/multilingual.less b/assets/less/multilingual.less
index c9d6a6d..d64c532 100644
--- a/assets/less/multilingual.less
+++ b/assets/less/multilingual.less
@@ -63,6 +63,8 @@
 
     &.field-multilingual-richeditor {
         .ml-btn {
+            width: 3rem;
+            padding-inline: 0 8px;
             border-radius: 0;
             border-bottom-left-radius: .375rem;
         }
@@ -74,12 +76,10 @@
     }
 
     &.field-multilingual-markdowneditor {
-        .editor-write {
-            margin-right: 2rem;
-        }
-
         .ml-btn {
             top: 41px;
+            width: 3rem;
+            padding-inline: 0 8px;
             border-radius: 0;
             border-bottom-left-radius: .375rem;
         }
diff --git a/formwidgets/mlricheditor/assets/js/mlricheditor.js b/formwidgets/mlricheditor/assets/js/mlricheditor.js
index e3d5c3e..f90dd2b 100644
--- a/formwidgets/mlricheditor/assets/js/mlricheditor.js
+++ b/formwidgets/mlricheditor/assets/js/mlricheditor.js
@@ -100,9 +100,9 @@
         var scrollHeight = $element.prop('scrollHeight') // int
         var offsetHeight = Math.round($element.outerHeight()); // round the float
         if (scrollHeight > offsetHeight) {
-            $element.css('padding-right', 40)
-            $btn.css('right', 16)
-            $dropdown.css('right', 14)
+            //$element.css('padding-right', 40)
+            //$btn.css('right', 16)
+            //$dropdown.css('right', 14)
         }
     }
```
![image](https://github.com/wintercms/wn-translate-plugin/assets/2013630/b9951376-bac6-4a9a-851f-1111833914dd)

mjauvin commented:
Ok, that version works best I think:
```diff
diff --git a/assets/less/multilingual.less b/assets/less/multilingual.less
index c9d6a6d..76b5350 100644
--- a/assets/less/multilingual.less
+++ b/assets/less/multilingual.less
@@ -63,6 +63,9 @@
 
     &.field-multilingual-richeditor {
         .ml-btn {
+            width: 4rem;
+            padding-inline: 0 1rem;
+            background: transparent;
             border-radius: 0;
             border-bottom-left-radius: .375rem;
         }
@@ -74,12 +77,10 @@
     }
 
     &.field-multilingual-markdowneditor {
-        .editor-write {
-            margin-right: 2rem;
-        }
-
         .ml-btn {
             top: 41px;
+            width: 3rem;
+            padding-inline: 0 8px;
             border-radius: 0;
             border-bottom-left-radius: .375rem;
         }
diff --git a/formwidgets/mlricheditor/assets/js/mlricheditor.js b/formwidgets/mlricheditor/assets/js/mlricheditor.js
index e3d5c3e..f90dd2b 100644
--- a/formwidgets/mlricheditor/assets/js/mlricheditor.js
+++ b/formwidgets/mlricheditor/assets/js/mlricheditor.js
@@ -100,9 +100,9 @@
         var scrollHeight = $element.prop('scrollHeight') // int
         var offsetHeight = Math.round($element.outerHeight()); // round the float
         if (scrollHeight > offsetHeight) {
-            $element.css('padding-right', 40)
-            $btn.css('right', 16)
-            $dropdown.css('right', 14)
+            //$element.css('padding-right', 40)
+            //$btn.css('right', 16)
+            //$dropdown.css('right', 14)
         }
     }
```
![image](https://github.com/wintercms/wn-translate-plugin/assets/2013630/d313062f-51c5-4232-abd3-5e9d8f6a5964)

AIC-BV commented:
I was looking into this as well
Due to the current HTML structure the button will always be above the scrollbar. The `.ml-btn` needs to be injected in `.fr-wrapper`. Then we can manipulate the z-index or the order of the HTML so that the content box is above the button.

Making it transparent removes consistency and you can't click the scrollbar below the button making it bad UX imo.
For me, your changes are fine since I don't click the scrollbar, but ideally the location of the button should be modified (if that is even possible at all) so we can do this the legit way

AIC-BV commented:
[How I feel right now](https://giphy.com/gifs/nfl-sports-football-sport-j4zjZdyqZZ7gsirl1M)

AIC-BV commented:
If all other options are not possible, then these are the only options:
1. Transparent button as you suggested
2. My latest commit, but hardcoded in CSS instead of JS. The button will ALWAYS have right: 16px; Regardless of the scrollbar, which make me think textarea will have this issue too?

mjauvin commented:
```diff
diff --git a/assets/less/multilingual.less b/assets/less/multilingual.less
index c9d6a6d..fcdb39a 100644
--- a/assets/less/multilingual.less
+++ b/assets/less/multilingual.less
@@ -17,7 +17,7 @@
         max-height: 38px;
 
         color: lighten(@multilingual-btn-color, 15%);
-        background-color: #F8F6F3;
+        background-color: #F8F6F380;
         box-shadow: none;
 
         font-size: 11px;
@@ -63,6 +63,7 @@
 
     &.field-multilingual-richeditor {
         .ml-btn {
+            right: 0.5rem;
             border-radius: 0;
             border-bottom-left-radius: .375rem;
         }
@@ -74,12 +75,10 @@
     }
 
     &.field-multilingual-markdowneditor {
-        .editor-write {
-            margin-right: 2rem;
-        }
-
         .ml-btn {
             top: 41px;
+            width: 3rem;
+            padding-inline: 0 8px;
             border-radius: 0;
             border-bottom-left-radius: .375rem;
         }
@@ -111,6 +110,11 @@
         }
     }
 }
+.dark .field-multilingual.field-multilingual-richeditor {
+    .ml-btn {
+        color: white;
+    }
+}
 
 // ?? CHECK THIS
 // .help-block.before-field + .field-multilingual.field-multilingual-textarea {
diff --git a/formwidgets/mlricheditor/assets/js/mlricheditor.js b/formwidgets/mlricheditor/assets/js/mlricheditor.js
index e3d5c3e..f90dd2b 100644
--- a/formwidgets/mlricheditor/assets/js/mlricheditor.js
+++ b/formwidgets/mlricheditor/assets/js/mlricheditor.js
@@ -100,9 +100,9 @@
         var scrollHeight = $element.prop('scrollHeight') // int
         var offsetHeight = Math.round($element.outerHeight()); // round the float
         if (scrollHeight > offsetHeight) {
-            $element.css('padding-right', 40)
-            $btn.css('right', 16)
-            $dropdown.css('right', 14)
+            //$element.css('padding-right', 40)
+            //$btn.css('right', 16)
+            //$dropdown.css('right', 14)
```
![image](https://github.com/wintercms/wn-translate-plugin/assets/2013630/1b46cf1c-4260-4780-8e71-1b6387e3037e)

mjauvin commented:
```diff
diff --git a/assets/less/multilingual.less b/assets/less/multilingual.less
index c9d6a6d..82e9e91 100644
--- a/assets/less/multilingual.less
+++ b/assets/less/multilingual.less
@@ -51,6 +51,8 @@
         }
         
         .ml-btn {
+            right: 0.5rem;
+            background: transparent;
             border-bottom-right-radius: 0px;
             border-bottom-left-radius: .375rem;
         }   
@@ -63,6 +65,7 @@
 
     &.field-multilingual-richeditor {
         .ml-btn {
+            right: 0.5rem;
             border-radius: 0;
             border-bottom-left-radius: .375rem;
         }   
@@ -74,12 +77,10 @@
     }
 
     &.field-multilingual-markdowneditor {
-        .editor-write {
-            margin-right: 2rem;
-        }
-        
         .ml-btn {
             top: 41px;
+            width: 3rem;
+            padding-inline: 0 8px;
             border-radius: 0;
             border-bottom-left-radius: .375rem;
         }
@@ -112,6 +113,15 @@
     }   
 }
             
+.dark {     
+    .field-multilingual :is(.field-multilingual-richeditor,.field-multilingual-textarea) {
+        .ml-btn {
+            background: transparent;
+            color: white;
+        }
+    }
+}
+
```

mjauvin commented:
```diff
diff --git a/assets/less/multilingual.less b/assets/less/multilingual.less
index c9d6a6d..ab0c038 100644
--- a/assets/less/multilingual.less
+++ b/assets/less/multilingual.less
@@ -51,6 +51,8 @@
         }
 
         .ml-btn {
+            right: 0.5rem;
+            background: transparent;
             border-bottom-right-radius: 0px;
             border-bottom-left-radius: .375rem;
         }
@@ -63,6 +65,8 @@
 
     &.field-multilingual-richeditor {
         .ml-btn {
+            right: 0.5rem;
+            background: transparent;
             border-radius: 0;
             border-bottom-left-radius: .375rem;
         }
@@ -74,12 +78,10 @@
     }
 
     &.field-multilingual-markdowneditor {
-        .editor-write {
-            margin-right: 2rem;
-        }
-
         .ml-btn {
             top: 41px;
+            width: 3rem;
+            padding-inline: 0 8px;
             border-radius: 0;
             border-bottom-left-radius: .375rem;
         }
@@ -112,6 +114,14 @@
     }
 }
 
+.dark {
+    .field-multilingual :is(.field-multilingual-richeditor,.field-multilingual-textarea) {
+        .ml-btn {
+            color: white;
+        }
+    }
+}
+
```

AIC-BV commented:
@mjauvin I think the current PR is my latest suggestion as discussed on Discord

mjauvin commented:
Works really well now! Nice work!
Can you do the same for textarea / markdown editor ?

Note: jQuery allows you to have multiple events for the `on()` method.
(e.g. `$element.on( "keyup keydown", setMLButtonPosition`)

AIC-BV commented:
Good thinking, I merged the two events!

AIC-BV commented:
TODO:
This functionality needs to be added to:
- textarea
- markdowneditor tab mode

AIC-BV commented:
Is there a global script running where I can add textarea code?
Or does it need an entire Class file like the other widgets have...?
```
const MLtextareas = document.querySelector('.field-multilingual-textarea .field-textarea)
for (const textarea of MLtextareas) {
    textarea.addEventListener('change', setMLButtonPosition)
}

function setMLButtonPosition() {
    // some checks for the scrollbar & update its position like in the richeditor field
}
```

AIC-BV commented:
> > Is there a global script running where I can add textarea code? Or does it need an entire Class file like the other widgets have...?
> > ```
> > const MLtextareas = document.querySelector('.field-multilingual-textarea .field-textarea)
> > for (const textarea of MLtextareas) {
> >     textarea.addEventListener('change', setMLButtonPosition)
> > }
> > 
> > function setMLButtonPosition() {
> >     // some checks for the scrollbar & update its position like in the richeditor field
> > }
> > ```
> 
> Let me think about it

Its like you said, it won't work if generated by repeater

AIC-BV commented:
Needs this https://github.com/wintercms/winter/pull/997 for markdowneditor

AIC-BV commented:
@mjauvin 😄

mjauvin commented:
Looks like it's working fine, but it's not triggering all the times when you get back from fullscreen mode (especially if you don't paste a huge amount of text).

AIC-BV commented:
> Looks like it's working fine, but it's not triggering all the times when you get back from fullscreen mode (especially if you don't paste a huge amount of text).

Can you show a video
For me it works every time but it snaps back with a delay

mjauvin commented:
@AIC-BV where were we on this?

AIC-BV commented:
For me it is finished 😄 
You had an issue you mentioned above but I can't reproduce it

mjauvin commented on formwidgets/mltextarea/partials/_mltextarea.htm:
```diff
@@ -15,6 +15,7 @@
             placeholder="<?= e(trans($field->placeholder)) ?>"
             class="form-control field-textarea size-<?= $field->size ?>"
             autocomplete="off"
+            onchange="updateTextareaLayout(this)"
```
```suggestion
            oninput="updateTextareaLayout(this)"
```

AIC-BV replied:
@mjauvin The reason why I used onchange instead of oninput is because onchange is triggered far less:
![image](https://github.com/wintercms/wn-translate-plugin/assets/89913092/280d46e1-b21d-4ead-b769-249b7c66256b)
![image](https://github.com/wintercms/wn-translate-plugin/assets/89913092/5f765cc2-ef91-4634-8ff2-abb75623055f)

The translation button is hidden when the element is focussed, so it only needs to update once and thats when you unfocus the element.
For me it doesn't matter wether its change or input but change looks more performant to me but I guess its just minor

mjauvin replied:
The button is not hidden if the TailwindUi plugin is not used.

mjauvin replied:
This should be the same with or without that plugin...

AIC-BV replied:
Z-index 10 on focus
![image](https://github.com/wintercms/wn-translate-plugin/assets/89913092/c3d23261-6339-4367-a19f-6392153995e0)

AIC-BV replied:
But oninput is the safer choice 👍

mjauvin commented:
@AIC-BV here's a video for the markdown editor that shows it is not updating:
[Screencast from 2024-03-18 12-48-53.webm](https://github.com/wintercms/wn-translate-plugin/assets/2013630/306f6329-5944-4702-b708-1e756b1b8ec3)

mjauvin commented:
I am using the Winter.Test plugin and added the following to the Post model:
```php
    public $implement = ['@Winter.Translate.Behaviors.TranslatableModel'];
    public $translatable = ['name', 'content_md', 'content_html'];
```

AIC-BV commented:
@mjauvin Markdowneditor has been fixed
